optimistix/_solver/quasi_newton.py

```diff
@@ -115,12 +78,615 @@ def __call__(
+        lambda x: x[(state.index_start - 1) % history_len],
+        (state.y_diff_history, state.grad_diff_history),
+    )
+    y_grad_diff_inner = tree_dot(latest_y_diff, latest_grad_diff)
+    grad_diff_norm_sq = tree_dot(latest_grad_diff, latest_grad_diff)
+    gamma_k = jnp.where(
+        grad_diff_norm_sq > 0, y_grad_diff_inner / grad_diff_norm_sq, 1.0
+    )
+    descent_direction = (gamma_k * descent_direction**ω).ω
+    (descent_direction, _), _ = jax.lax.scan(
+        forward_iter, (descent_direction, alpha), jnp.arange(history_len), reverse=False
+    )
+    return descent_direction
+
+
+def _lbfgs_hessian_operator_fn(
+    step: Y,  # TODO name this better?
+    # y_diff_history: PyTree[Y],
+    # grad_diff_history: PyTree[Y],
+    state: _LBFGSHessianUpdateState,
+) -> Y:
+    """Compact representation of the L-BFGS Hessian update.
+
+    Implements the L-BFGS update as described in Algorithm 3.2 of:
+
+        Byrd, R. H., Nocedal, J., & Schnabel, R. B. (1994).
+        "Representations of quasi-Newton matrices and their use in limited memory
+        methods." *Mathematical Programming*, 63(1), 129–156.
+
+    This representation is used in limited-memory BFGS algorithms to efficiently store
+    and apply approximate Hessians without forming full matrices.
+    """
+
+    history_len = jtu.tree_leaves(state.y_diff_history)[0].shape[0]
+
+    latest_grad_diff, latest_y_diff = jtu.tree_map(
+        lambda x: x[(state.index_start - 1) % history_len],
+        (state.grad_diff_history, state.y_diff_history),
+    )
+
+    curvature = tree_dot(latest_y_diff, latest_grad_diff)
+    gamma_k = jnp.where(
+        curvature > 0, tree_dot(latest_grad_diff, latest_grad_diff) / curvature, 1.0
+    )
+
+    # computation of eqn (3.22) in Byrd at al. 1994
+    # eqn (2.26)
+    safe_y_diff_grad_diff_inner = jnp.where(
+        state.y_diff_grad_diff_inner == 0, 1, state.y_diff_grad_diff_inner
+    )
+    J_square = gamma_k * state.y_diff_cross_inner + (
+        jnp.dot(
+            state.y_diff_grad_diff_cross_inner
+            / jnp.expand_dims(safe_y_diff_grad_diff_inner, axis=0),
+            state.y_diff_grad_diff_cross_inner.T,
+        )
+    )
+    J = jnp.linalg.cholesky(J_square)
+
+    # step 5 of algorithm 3.2 of Byrd at al. 1994
+    descent = jnp.hstack(
+        [
+            v_tree_dot(state.grad_diff_history, step),
+            gamma_k * v_tree_dot(state.y_diff_history, step),
+        ]
+    )
+
+    # sort history to match lower triangular structure.
+    roll_by = -(state.index_start % history_len)
+    conditional_sort = lambda array, roll: jax.lax.cond(
+        state.index_start <= history_len,
+        lambda x, r: x,
+        lambda x, r: jnp.roll(x, r),
+        array,
+        roll,
+    )
+    descent = descent.at[:history_len].set(
+        conditional_sort(descent[:history_len], roll_by)
+    )
+    descent = descent.at[history_len:].set(
+        conditional_sort(descent[history_len:], roll_by)
+    )
```

Comment:
Heads-up that a lot of this will likely be very computationally expensive in the same spirit as above. (Although I've not stared at it so closely as to determine whether a circular-buffer-like approach would suffice here.) In particular rolling always make a complete copy of the array.

If you already know all that and anyway think is the best way to implement this then feel free to ignore this comment!